The first error was definitely caused by my change.  I made an update and now I'm only seing:

```
ERROR: test_prog_w_funky_name (__main__.TestProcess)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_psutil.py", line 1802, in test_prog_w_funky_name
    self.assertEqual(p.cmdline(), cmdline)
  File "C:\Python27-x64\lib\site-packages\psutil-3.5.0-py2.7-win-amd64.egg\psutil\__init__.py", line 603, in cmdline
    return self._proc.cmdline()
  File "C:\Python27-x64\lib\site-packages\psutil-3.5.0-py2.7-win-amd64.egg\psutil\_pswindows.py", line 290, in wrapper
    raise NoSuchProcess(self.pid, self._name)
NoSuchProcess: psutil.NoSuchProcess process no longer exists (pid=1884, name='psuz8b8nufoo bar )')
```

https://ci.appveyor.com/project/giampaolo/psutil/build/370/job/khsp57dfqvvbomj1#L418

--- a/psutil/arch/windows/process_info.c
+++ b/psutil/arch/windows/process_info.c
@@ -203,10 +203,14 @@ handlep_is_running(HANDLE hProcess) {
     return 0;
 }
 
-/* Get the cmdline as a Python list of the process with the given pid.  On
-   success 0 is returned.  On error the given output parameters are not
+/* Get one or more parameters of the process with the given pid:
+
+   pcmdline: the command line as a Python list
+   pcwd: the current working directory as a Python string
+
+   On success 0 is returned.  On error the given output parameters are not
    touched, -1 is returned, and an appropriate Python exception is set. */
-static int psutil_get_parameters(long pid, PyObject **pcmdline) {
+static int psutil_get_parameters(long pid, PyObject **pcmdline, PyObject **pcwd) {
     int nArgs, i;
     LPWSTR *szArglist = NULL;
     HANDLE hProcess = NULL;
@@ -214,6 +218,8 @@ static int psutil_get_parameters(long pid, PyObject **pcmdline) {
     PVOID rtlUserProcParamsAddress;
     UNICODE_STRING commandLine;
     WCHAR *commandLineContents = NULL;
+    UNICODE_STRING currentDirectory;
+    WCHAR *currentDirectoryContent = NULL;
     PyObject *py_unicode = NULL;
     PyObject *py_retlist = NULL;
 
@@ -231,8 +237,13 @@ static int psutil_get_parameters(long pid, PyObject **pcmdline) {
                            &rtlUserProcParamsAddress, sizeof(PVOID), NULL))
 #endif
     {
-        ////printf("Could not read the address of ProcessParameters!\n");
-        PyErr_SetFromWindowsErr(0);
+        if (GetLastError() == ERROR_PARTIAL_COPY) {
+            // this occurs quite often with system processes
+            AccessDenied();
+        }
+        else {
+            PyErr_SetFromWindowsErr(0);
+        }
         goto error;
     }
 
@@ -296,7 +307,7 @@ static int psutil_get_parameters(long pid, PyObject **pcmdline) {
                     goto error;
                 if (PyList_Append(py_retlist, py_unicode))
                     goto error;
-                Py_XDECREF(py_unicode);
+                Py_CLEAR(py_unicode);
             }
         }
 
@@ -305,6 +316,70 @@ static int psutil_get_parameters(long pid, PyObject **pcmdline) {
         free(commandLineContents);
 
         *pcmdline = py_retlist;
+        py_retlist = NULL;
+    }
+
+    if (pcwd != NULL) {
+        // Read the currentDirectory UNICODE_STRING structure.
+        // 0x24 refers to "CurrentDirectoryPath" of RTL_USER_PROCESS_PARAMETERS
+        // structure, see:
+        // http://wj32.wordpress.com/2009/01/24/
+        //     howto-get-the-command-line-of-processes/
+#ifdef _WIN64
+        if (!ReadProcessMemory(hProcess, (PCHAR)rtlUserProcParamsAddress + 56,
+                               &currentDirectory, sizeof(currentDirectory), NULL))
+#else
+        if (!ReadProcessMemory(hProcess,
+                               (PCHAR)rtlUserProcParamsAddress + 0x24,
+                               &currentDirectory, sizeof(currentDirectory), NULL))
+#endif
+        {
+            if (GetLastError() == ERROR_PARTIAL_COPY) {
+                // this occurs quite often with system processes
+                AccessDenied();
+            }
+            else {
+                PyErr_SetFromWindowsErr(0);
+            }
+            goto error;
+        }
+
+        // allocate memory to hold cwd
+        currentDirectoryContent = (WCHAR *)malloc(currentDirectory.Length + 1);
+        if (currentDirectoryContent == NULL) {
+            PyErr_NoMemory();
+            goto error;
+        }
+
+        // read cwd
+        if (!ReadProcessMemory(hProcess, currentDirectory.Buffer,
+                               currentDirectoryContent, currentDirectory.Length,
+                               NULL))
+        {
+            if (GetLastError() == ERROR_PARTIAL_COPY) {
+                // this occurs quite often with system processes
+                AccessDenied();
+            }
+            else {
+                PyErr_SetFromWindowsErr(0);
+            }
+            goto error;
+        }
+
+        // null-terminate the string to prevent wcslen from returning
+        // incorrect length the length specifier is in characters, but
+        // currentDirectory.Length is in bytes
+        currentDirectoryContent[(currentDirectory.Length / sizeof(WCHAR))] = '\0';
+
+        // convert wchar array to a Python unicode string
+        py_unicode = PyUnicode_FromWideChar(
+            currentDirectoryContent, wcslen(currentDirectoryContent));
+        if (py_unicode == NULL)
+            goto error;
+        CloseHandle(hProcess);
+        free(currentDirectoryContent);
+        *pcwd = py_unicode;
+        py_unicode = NULL;
     }
 
     CloseHandle(hProcess);
@@ -320,6 +395,8 @@ error:
         free(commandLineContents);
     if (szArglist != NULL)
         LocalFree(szArglist);
+    if (currentDirectoryContent != NULL)
+        free(currentDirectoryContent);
     return -1;
 }
 
@@ -330,110 +407,14 @@ error:
 PyObject *
 psutil_get_cmdline(long pid) {
     PyObject *ret = NULL;
-    psutil_get_parameters(pid, &ret);
+    psutil_get_parameters(pid, &ret, NULL);
     return ret;
 }
 
 PyObject *psutil_get_cwd(long pid) {
-    HANDLE processHandle = NULL;
-    PVOID pebAddress;
-    PVOID rtlUserProcParamsAddress;
-    UNICODE_STRING currentDirectory;
-    WCHAR *currentDirectoryContent = NULL;
-    PyObject *py_unicode = NULL;
-
-    processHandle = psutil_handle_from_pid(pid);
-    if (processHandle == NULL)
-        return NULL;
-
-    pebAddress = psutil_get_peb_address(processHandle);
-
-    // get the address of ProcessParameters
-#ifdef _WIN64
-    if (!ReadProcessMemory(processHandle, (PCHAR)pebAddress + 32,
-                           &rtlUserProcParamsAddress, sizeof(PVOID), NULL))
-#else
-    if (!ReadProcessMemory(processHandle, (PCHAR)pebAddress + 0x10,
-                           &rtlUserProcParamsAddress, sizeof(PVOID), NULL))
-#endif
-    {
-        CloseHandle(processHandle);
-        if (GetLastError() == ERROR_PARTIAL_COPY) {
-            // this occurs quite often with system processes
-            return AccessDenied();
-        }
-        else {
-            return PyErr_SetFromWindowsErr(0);
-        }
-    }
-
-    // Read the currentDirectory UNICODE_STRING structure.
-    // 0x24 refers to "CurrentDirectoryPath" of RTL_USER_PROCESS_PARAMETERS
-    // structure, see:
-    // http://wj32.wordpress.com/2009/01/24/
-    //     howto-get-the-command-line-of-processes/
-#ifdef _WIN64
-    if (!ReadProcessMemory(processHandle, (PCHAR)rtlUserProcParamsAddress + 56,
-                           &currentDirectory, sizeof(currentDirectory), NULL))
-#else
-    if (!ReadProcessMemory(processHandle,
-                           (PCHAR)rtlUserProcParamsAddress + 0x24,
-                           &currentDirectory, sizeof(currentDirectory), NULL))
-#endif
-    {
-        CloseHandle(processHandle);
-        if (GetLastError() == ERROR_PARTIAL_COPY) {
-            // this occurs quite often with system processes
-            return AccessDenied();
-        }
-        else {
-            return PyErr_SetFromWindowsErr(0);
-        }
-    }
-
-    // allocate memory to hold cwd
-    currentDirectoryContent = (WCHAR *)malloc(currentDirectory.Length + 1);
-    if (currentDirectoryContent == NULL) {
-        PyErr_NoMemory();
-        goto error;
-    }
-
-    // read cwd
-    if (!ReadProcessMemory(processHandle, currentDirectory.Buffer,
-                           currentDirectoryContent, currentDirectory.Length,
-                           NULL))
-    {
-        if (GetLastError() == ERROR_PARTIAL_COPY) {
-            // this occurs quite often with system processes
-            AccessDenied();
-        }
-        else {
-            PyErr_SetFromWindowsErr(0);
-        }
-        goto error;
-    }
-
-    // null-terminate the string to prevent wcslen from returning
-    // incorrect length the length specifier is in characters, but
-    // currentDirectory.Length is in bytes
-    currentDirectoryContent[(currentDirectory.Length / sizeof(WCHAR))] = '\0';
-
-    // convert wchar array to a Python unicode string
-    py_unicode = PyUnicode_FromWideChar(
-        currentDirectoryContent, wcslen(currentDirectoryContent));
-    if (py_unicode == NULL)
-        goto error;
-    CloseHandle(processHandle);
-    free(currentDirectoryContent);
-    return py_unicode;
-
-error:
-    Py_XDECREF(py_unicode);
-    if (currentDirectoryContent != NULL)
-        free(currentDirectoryContent);
-    if (processHandle != NULL)
-        CloseHandle(processHandle);
-    return NULL;
+    PyObject *ret = NULL;
+    psutil_get_parameters(pid, NULL, &ret);
+    return ret;
 }
 
 

--- a/psutil/arch/windows/process_info.c
+++ b/psutil/arch/windows/process_info.c
@@ -203,13 +203,10 @@ handlep_is_running(HANDLE hProcess) {
     return 0;
 }
 
-
-/*
- * returns a Python list representing the arguments for the process
- * with given pid or NULL on error.
- */
-PyObject *
-psutil_get_cmdline(long pid) {
+/* Get the cmdline as a Python list of the process with the given pid.  On
+   success 0 is returned.  On error the given output parameters are not
+   touched, -1 is returned, and an appropriate Python exception is set. */
+static int psutil_get_parameters(long pid, PyObject **pcmdline) {
     int nArgs, i;
     LPWSTR *szArglist = NULL;
     HANDLE hProcess = NULL;
@@ -222,7 +219,7 @@ psutil_get_cmdline(long pid) {
 
     hProcess = psutil_handle_from_pid(pid);
     if (hProcess == NULL)
-        return NULL;
+        return -1;
     pebAddress = psutil_get_peb_address(hProcess);
 
     // get the address of ProcessParameters
@@ -239,69 +236,80 @@ psutil_get_cmdline(long pid) {
         goto error;
     }
 
-    // read the CommandLine UNICODE_STRING structure
+    if (pcmdline != NULL) {
+        // read the CommandLine UNICODE_STRING structure
 #ifdef _WIN64
-    if (!ReadProcessMemory(hProcess, (PCHAR)rtlUserProcParamsAddress + 112,
-                           &commandLine, sizeof(commandLine), NULL))
+        if (!ReadProcessMemory(hProcess, (PCHAR)rtlUserProcParamsAddress + 112,
+                               &commandLine, sizeof(commandLine), NULL))
 #else
-    if (!ReadProcessMemory(hProcess, (PCHAR)rtlUserProcParamsAddress + 0x40,
-                           &commandLine, sizeof(commandLine), NULL))
+        if (!ReadProcessMemory(hProcess, (PCHAR)rtlUserProcParamsAddress + 0x40,
+                               &commandLine, sizeof(commandLine), NULL))
 #endif
-    {
-        PyErr_SetFromWindowsErr(0);
-        goto error;
-    }
-
-
-    // allocate memory to hold the command line
-    commandLineContents = (WCHAR *)malloc(commandLine.Length + 1);
-    if (commandLineContents == NULL) {
-        PyErr_NoMemory();
-        goto error;
-    }
-
-    // read the command line
-    if (!ReadProcessMemory(hProcess, commandLine.Buffer,
-                           commandLineContents, commandLine.Length, NULL))
-    {
-        PyErr_SetFromWindowsErr(0);
-        goto error;
-    }
-
-    // Null-terminate the string to prevent wcslen from returning
-    // incorrect length the length specifier is in characters, but
-    // commandLine.Length is in bytes.
-    commandLineContents[(commandLine.Length / sizeof(WCHAR))] = '\0';
-
-    // attempt to parse the command line using Win32 API, fall back
-    // on string cmdline version otherwise
-    szArglist = CommandLineToArgvW(commandLineContents, &nArgs);
-    if (szArglist == NULL) {
-        PyErr_SetFromWindowsErr(0);
-        goto error;
-    }
-    else {
-        // arglist parsed as array of UNICODE_STRING, so convert each to
-        // Python string object and add to arg list
-        py_retlist = Py_BuildValue("[]");
-        if (py_retlist == NULL)
+        {
+            if (GetLastError() == ERROR_PARTIAL_COPY) {
+                // this occurs quite often with system processes
+                AccessDenied();
+            }
+            else {
+                PyErr_SetFromWindowsErr(0);
+            }
             goto error;
-        for (i = 0; i < nArgs; i++) {
-            py_unicode = PyUnicode_FromWideChar(
-                szArglist[i], wcslen(szArglist[i]));
-            if (py_unicode == NULL)
-                goto error;
-            if (PyList_Append(py_retlist, py_unicode))
-                goto error;
-            Py_XDECREF(py_unicode);
         }
+
+        // allocate memory to hold the command line
+        commandLineContents = (WCHAR *)malloc(commandLine.Length + 1);
+        if (commandLineContents == NULL) {
+            PyErr_NoMemory();
+            goto error;
+        }
+
+        // read the command line
+        if (!ReadProcessMemory(hProcess, commandLine.Buffer,
+                               commandLineContents, commandLine.Length, NULL))
+        {
+            PyErr_SetFromWindowsErr(0);
+            goto error;
+        }
+
+        // Null-terminate the string to prevent wcslen from returning
+        // incorrect length the length specifier is in characters, but
+        // commandLine.Length is in bytes.
+        commandLineContents[(commandLine.Length / sizeof(WCHAR))] = '\0';
+
+        // attempt to parse the command line using Win32 API, fall back
+        // on string cmdline version otherwise
+        szArglist = CommandLineToArgvW(commandLineContents, &nArgs);
+        if (szArglist == NULL) {
+            PyErr_SetFromWindowsErr(0);
+            goto error;
+        }
+        else {
+            // arglist parsed as array of UNICODE_STRING, so convert each to
+            // Python string object and add to arg list
+            py_retlist = Py_BuildValue("[]");
+            if (py_retlist == NULL)
+                goto error;
+            for (i = 0; i < nArgs; i++) {
+                py_unicode = PyUnicode_FromWideChar(
+                    szArglist[i], wcslen(szArglist[i]));
+                if (py_unicode == NULL)
+                    goto error;
+                if (PyList_Append(py_retlist, py_unicode))
+                    goto error;
+                Py_XDECREF(py_unicode);
+            }
+        }
+
+        if (szArglist != NULL)
+            LocalFree(szArglist);
+        free(commandLineContents);
+
+        *pcmdline = py_retlist;
     }
 
-    if (szArglist != NULL)
-        LocalFree(szArglist);
-    free(commandLineContents);
     CloseHandle(hProcess);
-    return py_retlist;
+
+    return 0;
 
 error:
     Py_XDECREF(py_unicode);
@@ -312,9 +320,19 @@ error:
         free(commandLineContents);
     if (szArglist != NULL)
         LocalFree(szArglist);
-    return NULL;
+    return -1;
 }
 
+/*
+ * returns a Python list representing the arguments for the process
+ * with given pid or NULL on error.
+ */
+PyObject *
+psutil_get_cmdline(long pid) {
+    PyObject *ret = NULL;
+    psutil_get_parameters(pid, &ret);
+    return ret;
+}
 
 PyObject *psutil_get_cwd(long pid) {
     HANDLE processHandle = NULL;

